File: backend/internal/infrastructure/rabbitmq.go
package infrastructure

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/amll-dev/amll-hub/backend/internal/config"
	amqp "github.com/rabbitmq/amqp091-go"
)

// RabbitMQ 封装连接与通道
type RabbitMQ struct {
	Conn    *amqp.Connection
	Channel *amqp.Channel
	Queue   amqp.Queue
	DLQ     amqp.Queue
	NFQueue amqp.Queue // 无歌词解析队列
	NFDLQ   amqp.Queue // 无歌词死信队列
	cfg     config.RabbitMQConfig
}

// 无歌词解析相关常量
const (
	NFExchange    = "ttml.not_found"
	NFRoutingKey  = "not_found.parse"
	NFDLXExchange = "ttml.not_found.dlx"
	NFDLQRouting  = "not_found.failed"
	NFQueueName   = "not_found_parse_queue"
	NFDLQName     = "not_found_parse_queue.dlq"
)

// NewRabbitMQ 初始化 RabbitMQ，声明主队列 + 死信队列
func NewRabbitMQ(cfg config.RabbitMQConfig) (*RabbitMQ, error) {
	conn, err := amqp.DialConfig(cfg.URL, amqp.Config{
		Heartbeat: 30 * time.Second,
		Locale:    "en_US",
	})
	if err != nil {
		return nil, fmt.Errorf("dial rabbitmq: %w", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("open channel: %w", err)
	}

	// 声明死信交换机与队列
	dlxName := "ttml.sync.dlx"
	if err := ch.ExchangeDeclare(
		dlxName, "direct", true, false, false, false, nil,
	); err != nil {
		return nil, fmt.Errorf("declare dlx exchange: %w", err)
	}

	dlq, err := ch.QueueDeclare(
		cfg.DLQ, true, false, false, false, nil,
	)
	if err != nil {
		return nil, fmt.Errorf("declare dlq: %w", err)
	}
	if err := ch.QueueBind(dlq.Name, "sync.failed", dlxName, false, nil); err != nil {
		return nil, fmt.Errorf("bind dlq: %w", err)
	}

	// 声明主交换机与队列（绑定 DLX）
	exName := "ttml.sync"
	if err := ch.ExchangeDeclare(
		exName, "direct", true, false, false, false, nil,
	); err != nil {
		return nil, fmt.Errorf("declare exchange: %w", err)
	}

	queue, err := ch.QueueDeclare(
		cfg.Queue, true, false, false, false,
		amqp.Table{
			"x-dead-letter-exchange":    dlxName,
			"x-dead-letter-routing-key": "sync.failed",
		},
	)
	if err != nil {
		return nil, fmt.Errorf("declare queue: %w", err)
	}
	if err := ch.QueueBind(queue.Name, "sync.request", exName, false, nil); err != nil {
		return nil, fmt.Errorf("bind queue: %w", err)
	}

	// QoS：串行消费
	if err := ch.Qos(1, 0, false); err != nil {
		return nil, fmt.Errorf("qos: %w", err)
	}

	// === 无歌词解析队列（独立交换机/队列/DLQ） ===
	nfDlxName := NFDLXExchange
	if err := ch.ExchangeDeclare(
		nfDlxName, "direct", true, false, false, false, nil,
	); err != nil {
		return nil, fmt.Errorf("declare nf dlx exchange: %w", err)
	}

	nfDlq, err := ch.QueueDeclare(
		NFDLQName, true, false, false, false, nil,
	)
	if err != nil {
		return nil, fmt.Errorf("declare nf dlq: %w", err)
	}
	if err := ch.QueueBind(nfDlq.Name, NFDLQRouting, nfDlxName, false, nil); err != nil {
		return nil, fmt.Errorf("bind nf dlq: %w", err)
	}

	nfExName := NFExchange
	if err := ch.ExchangeDeclare(
		nfExName, "direct", true, false, false, false, nil,
	); err != nil {
		return nil, fmt.Errorf("declare nf exchange: %w", err)
	}

	nfQueue, err := ch.QueueDeclare(
		NFQueueName, true, false, false, false,
		amqp.Table{
			"x-dead-letter-exchange":    nfDlxName,
			"x-dead-letter-routing-key": NFDLQRouting,
		},
	)
	if err != nil {
		return nil, fmt.Errorf("declare nf queue: %w", err)
	}
	if err := ch.QueueBind(nfQueue.Name, NFRoutingKey, nfExName, false, nil); err != nil {
		return nil, fmt.Errorf("bind nf queue: %w", err)
	}

	return &RabbitMQ{
		Conn:    conn,
		Channel: ch,
		Queue:   queue,
		DLQ:     dlq,
		NFQueue: nfQueue,
		NFDLQ:   nfDlq,
		cfg:     cfg,
	}, nil
}

// PublishSyncRequest 发布一条同步任务到主队列
func (r *RabbitMQ) PublishSyncRequest(body []byte, requestID string, triggeredBy string) error {
	return r.Channel.Publish(
		"ttml.sync",
		"sync.request",
		false,
		false,
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			MessageId:    requestID,
			Timestamp:    time.Now(),
			Headers: amqp.Table{
				"x-request-id":   requestID,
				"x-triggered-by": triggeredBy,
			},
			Body: body,
		},
	)
}

// NotFoundParseMessage 无歌词解析消息体
type NotFoundParseMessage struct {
	Platform   string `json:"platform"`
	PlatformID string `json:"platformId"`
	ClientIP   string `json:"clientIp"`
}

// PublishNotFoundParse 发布一条无歌词解析任务
func (r *RabbitMQ) PublishNotFoundParse(msg NotFoundParseMessage) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("marshal nf message: %w", err)
	}
	return r.Channel.Publish(
		NFExchange,
		NFRoutingKey,
		false,
		false,
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now(),
			MessageId:    fmt.Sprintf("%s:%s", msg.Platform, msg.PlatformID),
			Body:         body,
		},
	)
}

// Close 关闭连接
func (r *RabbitMQ) Close() error {
	if r.Channel != nil {
		_ = r.Channel.Close()
	}
	if r.Conn != nil {
		return r.Conn.Close()
	}
	return nil
}

File: worker/src/infra/rabbitmq.rs
use anyhow::{Context, Result};
use lapin::{
    options::{BasicQosOptions, ExchangeDeclareOptions, QueueBindOptions, QueueDeclareOptions},
    types::{AMQPValue, FieldTable},
    Channel, Connection, ConnectionProperties, ExchangeKind,
};

use crate::config::Config;

/// 初始化 RabbitMQ 连接与队列声明
pub struct RabbitMq {
    #[allow(dead_code)]
    pub conn: Connection,
    pub channel: Channel,
}

impl RabbitMq {
    #[allow(dead_code)]
    pub async fn channel(&self) -> Result<Channel> {
        let ch = self.conn.create_channel().await.context("create channel")?;
        Ok(ch)
    }
}

pub async fn init_rabbitmq(cfg: &Config) -> Result<RabbitMq> {
    let conn = Connection::connect(&cfg.rabbitmq.url, ConnectionProperties::default())
        .await
        .context("connect rabbitmq")?;

    let channel = conn.create_channel().await.context("create channel")?;

    // DLX exchange + queue
    let dlx = "ttml.sync.dlx";
    channel
        .exchange_declare(
            dlx,
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare dlx exchange")?;

    channel
        .queue_declare(
            &cfg.rabbitmq.dlq,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare dlq")?;
    channel
        .queue_bind(
            &cfg.rabbitmq.dlq,
            dlx,
            "sync.failed",
            QueueBindOptions::default(),
            FieldTable::default(),
        )
        .await
        .context("bind dlq")?;

    // 主 exchange + queue（绑定 DLX）
    let ex = "ttml.sync";
    channel
        .exchange_declare(
            ex,
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare exchange")?;

    let mut args = FieldTable::default();
    args.insert(
        "x-dead-letter-exchange".into(),
        AMQPValue::LongString("ttml.sync.dlx".into()),
    );
    args.insert(
        "x-dead-letter-routing-key".into(),
        AMQPValue::LongString("sync.failed".into()),
    );
    channel
        .queue_declare(
            &cfg.rabbitmq.queue,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            args,
        )
        .await
        .context("declare queue")?;
    channel
        .queue_bind(
            &cfg.rabbitmq.queue,
            ex,
            "sync.request",
            QueueBindOptions::default(),
            FieldTable::default(),
        )
        .await
        .context("bind queue")?;

    channel
        .basic_qos(1, BasicQosOptions { global: false })
        .await
        .context("qos")?;

    // === not_found 解析队列（独立交换机/队列/DLQ） ===
    let nf_dlx = "ttml.not_found.dlx";
    channel
        .exchange_declare(
            nf_dlx,
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare nf dlx exchange")?;

    channel
        .queue_declare(
            &cfg.rabbitmq.nf_dlq,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare nf dlq")?;
    channel
        .queue_bind(
            &cfg.rabbitmq.nf_dlq,
            nf_dlx,
            "not_found.failed",
            QueueBindOptions::default(),
            FieldTable::default(),
        )
        .await
        .context("bind nf dlq")?;

    let nf_ex = "ttml.not_found";
    channel
        .exchange_declare(
            nf_ex,
            ExchangeKind::Direct,
            ExchangeDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await
        .context("declare nf exchange")?;

    let mut nf_args = FieldTable::default();
    nf_args.insert(
        "x-dead-letter-exchange".into(),
        AMQPValue::LongString(nf_dlx.into()),
    );
    nf_args.insert(
        "x-dead-letter-routing-key".into(),
        AMQPValue::LongString("not_found.failed".into()),
    );
    channel
        .queue_declare(
            &cfg.rabbitmq.nf_queue,
            QueueDeclareOptions {
                durable: true,
                ..Default::default()
            },
            nf_args,
        )
        .await
        .context("declare nf queue")?;
    channel
        .queue_bind(
            &cfg.rabbitmq.nf_queue,
            nf_ex,
            "not_found.parse",
            QueueBindOptions::default(),
            FieldTable::default(),
        )
        .await
        .context("bind nf queue")?;

    Ok(RabbitMq { conn, channel })
}
